### PR TITLE
Paper: Add updates for gate modifiers

### PR DESCRIPTION
**Pull request by zachschoenfeld33**
### Summary

Add additional spec details and include new concepts from the paper https://arxiv.org/pdf/2104.14722.pdf for the gate modifiers `ctrl/negctrl/pow/inv`.

The `stdgates.inc` library file is also updated from the paper.

### CODE REVIEW

File: examples/stdgates.inc
Comment (zachschoenfeld33) [inline examples/stdgates.inc:1]: This is copied exactly from pp. 13,14 of https://arxiv.org/pdf/2104.14722.pdf

File: source/grammar/tests/outputs/gate_modifiers.yaml
Comment (zachschoenfeld33) [inline source/grammar/tests/outputs/gate_modifiers.yaml:3]: Moved the empty gate test here.

File: source/language/gates.rst
Comment (zachschoenfeld33) [inline source/language/gates.rst:364]: Should the mathematical definition of power in terms of principal logarithm be added here? (See pp.16 of https://arxiv.org/pdf/2104.14722.pdf) It might be more intuitive to define the `kth` power by raising each eigenvalue to the `kth` power. That is, is we diagonalize the unitary as `U=SDS*` (for unitary `S` and diagonal matrix `D`), then `U^k = SD^kS*`

File: source/language/gates.rst
Comment (blakejohnson) [inline source/language/gates.rst:167]: I see that you need that in order to parse the `p` gate in `stdgates.inc`. But, I wonder if that wasn't actually a typo, as I am not sure what it means to apply a global phase to a qubit.

File: source/language/gates.rst
Comment (blakejohnson) [inline source/language/gates.rst:295]: I don't think controlled phase is a special case. In `stdgates.inc` we have:
```
gate cphase(lambda) a, b { ctrl @ phase(lambda) a, b; }
```
which properly resolves to a 2-qubit gate.

File: source/language/gates.rst
Comment (blakejohnson) [inline source/language/gates.rst:302]: The way this is written in `stdgates` is clearer, i.e. `ctrl @ rz(θ) q1, q2;`.

File: source/language/gates.rst
Comment (blakejohnson) [inline source/language/gates.rst:362]: For explanatory purposes it would be nice to show that this can also be written as `rz(-θ) q1`.

File: source/language/gates.rst
Comment (blakejohnson) [reply source/language/gates.rst:364]: @ajavadia or @awcross1 have an opinion here?

**Review by blakejohnson — CHANGES_REQUESTED** [review]
Thanks, Zach. I think the only real issue is whether `gphase` can take a qubit argument. I marked a few other things as well.

File: source/language/gates.rst
Comment (zachschoenfeld33) [reply source/language/gates.rst:295]: This is copied exactly from pp. 14 of https://arxiv.org/pdf/2104.14722.pdf. In "The control modifier" paragraph. Happy to remove though

Comment (blakejohnson) [reply source/language/gates.rst:295]: That sentence in the paper appears to define a 1-qubit gate, not a 2-qubit `cphase`. @awcross1 can you confirm?

Comment (zachschoenfeld33) [reply source/language/gates.rst:295]: Ah. Ok I think I interpreted it wrong. `ctrl @ gphase(lambda)` is just the 1q phase gate (although a bit odd bc control using implies two qubits).

Comment (zachschoenfeld33) [reply source/language/gates.rst:295]: That's why I thought it was a "special case"

Comment (awcross1) [reply source/language/gates.rst:295]: That's right, it's a 1-qubit gate. The paper should say "controlled *global* phase gate" to avoid the confusion with the cphase gate.

File: source/language/gates.rst
Comment (awcross1) [reply source/language/gates.rst:364]: The two definitions are essentially the same except that the principal logarithm restricts the arguments of the elements of D to be in (-pi,pi] so that the logarithm is unique. @ajavadia

File: source/language/gates.rst
Comment (zachschoenfeld33) [reply source/language/gates.rst:367]: Addressed in https://github.com/Qiskit/openqasm/pull/203/commits/4471686f9e8f72a249b267a99622e1a86150777b.

File: source/language/gates.rst
Comment (zachschoenfeld33) [reply source/language/gates.rst]: Addressed in https://github.com/Qiskit/openqasm/pull/203/commits/4471686f9e8f72a249b267a99622e1a86150777b.

File: source/language/gates.rst
Comment (zachschoenfeld33) [reply source/language/gates.rst]: Addressed in https://github.com/Qiskit/openqasm/pull/203/commits/4471686f9e8f72a249b267a99622e1a86150777b.

File: source/language/gates.rst
Comment (zachschoenfeld33) [reply source/language/gates.rst]: Addressed via controlled global phase gate language. See https://github.com/Qiskit/openqasm/pull/203/commits/4471686f9e8f72a249b267a99622e1a86150777b.

**Review by blakejohnson — DISMISSED** [review]

File: source/language/gates.rst
Comment (blakejohnson) [reply source/language/gates.rst:373]: Looks like we can go either way. I am fine to merge as is, but if you prefer to add this, let me know and I will hold off on clicking 'merge'.

Comment (zachschoenfeld33) [reply source/language/gates.rst:373]: I am also fine as is, but if we want it changed I'm happy to do so

Comment (awcross1) [reply source/language/gates.rst:373]: I'm fine with it as is except we might need to say "... for some integer or floating point ..." since it doesn't need to be positive.

**Review by blakejohnson — APPROVED** [review]

**Pull request merged**
